### ISSUE TITLE
NaNs in Model

### ISSUE THREAD

**Issue by inkrement** [benfred/implicit#376]
I am using the LogisticMatrixFactorization and in some cases it exhibits numerical instabilities. While training it produces NaNs, which spread over the full model. Interestingly enough, this happens to me quite often. More than half of all models end in NaN matrices.

**Comment by mohjeih** [comment]
I'm also having the same problem. Have you found a solution?

**Comment by inkrement** [comment]
I am still investigating this issue, which is not trivial at all: Although, the package (GitHub version) allows to set a seed for the random number generator, the results change from run to run. Therefore, I cannot reproduce results properly, which makes it quite challenging. I guess this has to do with multi-threading, therefore, I'll try next to set the number of threads to 1... let's see if it helps. 

Regarding the NaN-issue itself: I guess it is related to several factors. First of all, the package uses '03' as well as 'ffast-math' compiler flags. The first one enables quite eager optimisations, while the later one relaxes many assumptions and disables important checks. IMHO this combination is already quite risky (if I remember it correctly, even the gcc documentation mentions that you should not do so). Secondly, the default learning rate of 1 seems quite high for me. Especially, in combination with a logistic loss function, I would go for a smaller one. Both changes seem to help a bit, but NaNs still occur. Finally, we should investigate all divisions - NaNs often occur when you try to divide by zero.

**Comment by ita9naiwa** [comment]
I followed the paper author's implementation default parameter https://github.com/MrChrisJohnson/logistic-mf/blob/90d4d7d44fcd0f373fbb5a827c2dc9f5da68d84b/logistic_mf.py#L28

Btw, it's quite unstable.


I think this is related to two things..

1. 
for Adagrad optimizer, its usual to set a default learning rate to be 1.0, but this brings unstable behavior, it would be better to reduce the default learning rate to be between 0.1 - 0.3.


2. 
derivation of log-sigmoid function is quite unstable. especially, my implementation of this is quite numerically unstable.
we can add capping or replace the derivation calculation with robust ones.

**Comment by inkrement** [comment]
Thanks @ita9naiwa for the sigmoid hint. I replaced it with a more stable version. Nevertheless, the main issue in my case was indeed related to a division by 0 (or at least a very small number). In particular, adagrad scales the learning rate x derivative by the square root of the accumulated squared gradients. I just added a small constant to this value as it is done by most other implementation and now it works fine (at least in my cases). 

I'll provide a pull request in the next seconds, can you take a look at it?

**Comment by benfred** [comment]
Closed w/ #383
